### PR TITLE
feat: OwnerProxy

### PR DESCRIPTION
**Pull request by maximebrugel**
<div align="center">image</div>
<br>

In order to complete the ownership architecture, we need the `OwnerProxy` contract in charge of executing scripts for the Timelock (run transactions atomically).

To show how the contract works, I added two scripts : 
- `AddOperator.sol`, to add an operator (import, add and rebuild cache)
- `UpdateFees`, to update the entry and exit fees.

We can also use a single OwnerProxy, but it must use `upgradeToAndCall` to call the NestedFactory owner functions :

<div align="center">image</div>

### CODE REVIEW

File: test/unit/OwnerProxy.unit.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                0,
+                80,
+            ]);
+
+            await expect(
+                ownerProxy.connect(context.masterDeployer).execute(scriptUpdateFees.address, wrongScriptCalldata),
+            ).to.be.revertedWith("NF: ZERO_FEES");
+        });
+
+        it("Cant update fees if wrong exit fees is zero", async () => {
```
Comment (Yashiru) [inline test/unit/OwnerProxy.unit.ts:100]:
```suggestion
        it("Cant update fees if exit fees are zero", async () => {
```

File: test/unit/OwnerProxy.unit.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+            await expect(
+                ownerProxy
+                    .connect(context.masterDeployer)
+                    .execute(ethers.constants.AddressZero, scriptUpdateFeesCalldata),
+            ).to.be.revertedWith("OP: INVALID_TARGET");
+        });
+    });
+
+    describe("Update fees", () => {
+        it("Cant update fees if wrong entry fees is zero", async () => {
```
Comment (Yashiru) [inline test/unit/OwnerProxy.unit.ts:87]:
```suggestion
        it("Cant update fees if entry fees are zero", async () => {
```

File: test/unit/OwnerProxy.unit.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            await ownerProxy
+                .connect(context.masterDeployer)
+                .execute(scriptUpdateFees.address, scriptUpdateFeesCalldata);
+
+            expect(await context.nestedFactory.entryFees()).to.be.equal(BigNumber.from(30));
+            expect(await context.nestedFactory.exitFees()).to.be.equal(BigNumber.from(80));
+        });
+    });
+
+    describe("Add operator", () => {
```
Comment (Yashiru) [inline test/unit/OwnerProxy.unit.ts:129]: Add a test that should revert if the calldata NestedFactory address is zero

File: contracts/governance/scripts/AddOperator.sol [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    /// @notice Call NestedFactory and OperatorResolver to add an operator.
+    /// @param nestedFactory The NestedFactory address
+    /// @param operator The operator to add
+    /// @param name The operator bytes32 name
+    function addOperator(
+        address nestedFactory,
+        IOperatorResolver.Operator memory operator,
+        bytes32 name
+    ) external {
+        IOperatorResolver resolver = IOperatorResolver(MixinOperatorResolver(nestedFactory).resolver());
```
Comment (Yashiru) [inline contracts/governance/scripts/AddOperator.sol:18]: In case the address of the NestedFactory is not correct, wouldn't it be cheaper in gas to make a **_require_** to check it first?

File: contracts/governance/OwnerProxy.sol [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+/// @title Owner proxy to run atomic actions.
+/// @notice DSProxy-like contract without a cache to simply run
+///         a sequence of atomic actions.
+contract OwnerProxy is Ownable {
+    /// @notice Execute atomic actions. Only the owner can call this function (e.g. the timelock)
+    /// @param _target Address of the "script" to perform a delegatecall
+    /// @param _data The bytes calldata (bytes4 selector must be included)
```
Comment (adrien-supizet) [inline contracts/governance/OwnerProxy.sol:12]: `(bytes4 selector must be included)` should be a `/// @dev` imo

File: contracts/governance/scripts/AddOperator.sol [hunk 6]
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.11;
+
+import "../../interfaces/INestedFactory.sol";
+import "../../interfaces/IOperatorResolver.sol";
+import "../../abstracts/MixinOperatorResolver.sol";
+
+contract AddOperator {
```
Comment (adrien-supizet) [inline contracts/governance/scripts/AddOperator.sol:8]: Could be great to also have the RemoveOperator ready to prevent making mistakes if we ever need to do it urgently.

**Review by adrien-supizet — APPROVED** [review]

File: test/unit/OwnerProxy.unit.ts [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    });
+
+    describe("remove operator", () => {
+        beforeEach("Add flatTest operator to be removed", async () => {
+            await ownerProxy
+                .connect(context.masterDeployer)
+                .execute(scriptAddOperator.address, scriptAddOperatorCalldata);
+        });
+        
+        it("Cant remove operator if nested factory address is zero", async () => {
```
Comment (obatirou) [inline test/unit/OwnerProxy.unit.ts:217]:
```suggestion
        it("Cannot remove operator if nested factory address is zero", async () => {
```
or 
```suggestion
        it("Can't remove operator if nested factory address is zero", async () => {
```

File: contracts/governance/scripts/DeployAddOperator.sol [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        address deployedAddress;
+        assembly {
+            deployedAddress := create(0, add(bytecode, 0x20), mload(bytecode))
+        }
+        require(deployedAddress != address(0), "DAO-SCRIPT: FAILED_DEPLOY");
+
+        IOperatorResolver resolver = IOperatorResolver(MixinOperatorResolver(nestedFactory).resolver());
+
+        // Init arrays
+        uint256 operatorLength = operators.length;
```
Comment (Yashiru) [inline contracts/governance/scripts/DeployAddOperator.sol:36]:
```suggestion
        uint256 operatorLength = operators.length;
        require(operatorLength != 0, "DAO-SCRIPT: INVALID_OPERATOR_LEN");
```

File: contracts/governance/scripts/DeployAddOperator.sol [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        bytes32[] memory names = new bytes32[](operatorLength);
+        IOperatorResolver.Operator[] memory operatorsToImport = new IOperatorResolver.Operator[](operatorLength);
+
+        for (uint256 i; i < operatorLength; i++) {
+            IOperatorResolver.Operator memory operator = IOperatorResolver.Operator(
+                deployedAddress,
+                operators[i].selector
+            );
+            names[i] = operators[i].name;
+            operatorsToImport[i] = operator;
```
Comment (Yashiru) [inline contracts/governance/scripts/DeployAddOperator.sol:46]: Maybe store the new IOperatorResolver.Operator directly in the operatorsToImport array instead of storing it in a memory cached variable can save some gas.
 
```suggestion
            operatorsToImport[i] = IOperatorResolver.Operator(
                deployedAddress,
                operators[i].selector
            );;
```

File: contracts/governance/scripts/DeployAddOperator.sol [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+        // Start importing operators
+        IOperatorResolver(resolver).importOperators(names, operatorsToImport, destinations);
+
+        // Add all the operators to the factory
+        for (uint256 i; i < operatorLength; i++) {
+            INestedFactory(nestedFactory).addOperator(operators[i].name);
+        }
+
+        require(MixinOperatorResolver(nestedFactory).isResolverCached(), "DAO-SCRIPT: UNRESOLVED_CACHE");
```
Comment (Yashiru) [inline contracts/governance/scripts/DeployAddOperator.sol:61]: I am not very familiar with the MixinOperatorResolver, I wonder if this line is absolutely mandatory?

File: contracts/governance/scripts/DeployAddOperator.sol [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+    /// @notice Deploy and add operators
+    /// @dev One address and multiple selectors/names
+    /// @param nestedFactory The NestedFactory address
+    /// @param bytecode Operator implementation bytecode
+    /// @param operators Array of tuples => bytes32/bytes4 (name and selector)
+    function deployAddOperator(
+        address nestedFactory,
+        bytes memory bytecode,
+        tupleOperator[] memory operators
```
Comment (adrien-supizet) [inline contracts/governance/scripts/DeployAddOperator.sol:22]: It's a bit confusing:
- method name is singular
- one bytecode as param

but:

- `operators` param as an array - plural

Consider renaming `operators`

Comment (maximebrugel) [reply contracts/governance/scripts/DeployAddOperator.sol:22]: In the `OperatorResolver`, an Operator is an `address` (implementation) and a `bytes4` (selector), linked to a `bytes32` (name).

The `bytecode` param will be deployed and as a result, the implementation (address). 
So, the `operators` is the array with `name` and the `selector`, it is really an operator list to add.

I don't know how we could rename `operators`. Any idea ?

File: contracts/governance/scripts/DeployAddOperator.sol (hunk 10 above)
Comment (maximebrugel) [reply contracts/governance/scripts/DeployAddOperator.sol:61]: Not mandatory, but good to have

File: contracts/governance/scripts/DeployAddOperator.sol (hunk 11 above)
Comment (obatirou) [reply contracts/governance/scripts/DeployAddOperator.sol:22]: I would remove tuple from the `tupleOperator` name, it is explicit by looking at the type, no ?
Maybe name the struct `operatorFunction` and the array `operatorFunctions`  ?

Comment (adrien-supizet) [reply contracts/governance/scripts/DeployAddOperator.sol:22]: I understand how it works, then the method name is inaccurate, and should be plural.

Either the operator is the contract, or it's the address + selector + bytes. It just needs to be consistent across the protocol

Comment (maximebrugel) [reply contracts/governance/scripts/DeployAddOperator.sol:22]: @adrien-supizet I misunderstood, I'm updating to `DeployAndAddOperators`

File: contracts/governance/scripts/OperatorScripts.sol [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    /// @param name The operator bytes32 name
+    function removeOperator(bytes32 name) external {
+        ITransparentUpgradeableProxy(nestedFactory).upgradeToAndCall(
+            ITransparentUpgradeableProxy(nestedFactory).implementation(),
+            abi.encodeWithSelector(INestedFactory.removeOperator.selector, name)
+        );
+
+        // Init arrays with length 1 (only one operator to remove)
+        bytes32[] memory names = new bytes32[](1);
+        IOperatorResolver.Operator[] memory operatorsToImport = new IOperatorResolver.Operator[](1);
```
Comment (adrien-supizet) [inline contracts/governance/scripts/OperatorScripts.sol:98]: `operatorsToRemove`?

File: contracts/governance/scripts/OperatorScripts.sol [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        MixinOperatorResolver[] memory destinations = new MixinOperatorResolver[](1);
+
+        names[0] = name;
+        operatorsToImport[0] = IOperatorResolver.Operator({ implementation: address(0), selector: bytes4(0) });
+        destinations[0] = MixinOperatorResolver(nestedFactory);
+
+        IOperatorResolver(resolver).importOperators(names, operatorsToImport, destinations);
```
Comment (adrien-supizet) [inline contracts/governance/scripts/OperatorScripts.sol:105]: `importOperators` It does not look like we're removing an operator here. Am I missing something?

Comment (adrien-supizet) [reply contracts/governance/scripts/OperatorScripts.sol:105]: I'm missing something here, I'll take a look at NestedFactory's removeOperator method before finishing the review

Comment (adrien-supizet) [reply contracts/governance/scripts/OperatorScripts.sol:105]: Ok, I think the OperatorResolver should have had a removeOperators method, but this works too.

**Review by adrien-supizet — APPROVED** [review]
